[DOC-257] [DOC-289] [2.21] Add Catalog flags to yb-tserver reference page (#21369)
yugabyte/yugabyte-db: docs/content/preview/reference/configuration/yb-tserver.md

Default: `1GB`

## Catalog flags

##### ysql_catalog_preload_additional_table_list

> Specifies the names of catalog tables (such as pg_operator, pg_proc, and pg_amop) to be preloaded by PostgreSQL backend processes. This flag reduces latency of first execution of a particular statement on a connection.

Reviewer: > such as `pg_operator`, `pg_proc`, and `pg_amop`

To keep consistency with the rest of the docs, let's put the table names in backticks.

> Specifies the names of catalog tables (such as pg_operator, pg_proc, and pg_amop) to be preloaded by PostgreSQL backend processes. This flag reduces latency of first execution of a particular statement on a connection.

Reviewer: This flag reduces latency of first execution of a particular statement on a connection. -> This flag reduces the first query execution latency on new connections.

Default: `""`

If [ysql_catalog_preload_additional_tables](#ysql-catalog-preload-additional-tables) is also specified, the union of the above specified catalog tables and pg_am, pg_amproc, pg_cast, and pg_tablespace is preloaded.

##### ysql_catalog_preload_additional_tables

When enabled, PG backend processes preloads the catalog tables: `pg_am`, `pg_amproc`, `pg_cast`, and `pg_tablespace`. This flag reduces latency of first execution of a particular statement on a connection.

> Default: false

Reviewer: Should the default values be in backticks (e.g. `false`)? This seems to be the convention based on the [existing flag docs](https://docs.yugabyte.com/preview/reference/configuration/yb-tserver/#stop-logging-if-full-disk).

If [ysql_catalog_preload_additional_table_list](#ysql-catalog-preload-additional-table-list) is also specified, the union of `pg_am`, `pg_amproc`, `pg_cast`, and `pg_tablespace` and the tables specified in `ysql_catalog_preload_additional_table_list` is preloaded.

##### ysql_enable_read_request_caching

Enables the YB-TServer catalog cache, which reduces YB-Master overhead for starting a connection and internal system catalog metadata refresh (for example, after executing a DDL), when there are many YSQL connections per node.

Default: true

##### ysql_use_relcache_file

> Controls whether to utilize the PostgreSQL relcache init file, which caches critical system catalog entries. If enabled, each PostgreSQL connection loads only these minimal set of cached entries (except if the relcache init file needs to be re-built, for example, after a DDL invalidates the cache). If disabled, each PostgreSQL connection preloads the catalog cache, which consumes more memory but reduced first query latency.

Reviewer: but reduced first query latency -> but reduces first query latency

> Controls whether to utilize the PostgreSQL relcache init file, which caches critical system catalog entries. If enabled, each PostgreSQL connection loads only these minimal set of cached entries (except if the relcache init file needs to be re-built, for example, after a DDL invalidates the cache). If disabled, each PostgreSQL connection preloads the catalog cache, which consumes more memory but reduced first query latency.

Reviewer: each PostgreSQL connection loads only these minimal set -> each PostgreSQL connection only loads this minimal set

Default: `read committed`

> ## Catalog flags

Reviewer 2: Why put catalog flags and advanced flags after PG server options? I would keep flags together and put options last
https://deploy-preview-21369--infallible-bardeen-164bc9.netlify.app/preview/reference/configuration/yb-tserver/#postgresql-server-options

##### ysql_catalog_preload_additional_table_list

Specifies the names of catalog tables (such as `pg_operator`, `pg_proc`, and `pg_amop`) to be preloaded by PostgreSQL backend processes. This flag reduces latency of first query execution of a particular statement on a connection.

Default: `""`

If [ysql_catalog_preload_additional_tables](#ysql-catalog-preload-additional-tables) is also specified, the union of the above specified catalog tables and `pg_am`, `pg_amproc`, `pg_cast`, and `pg_tablespace` is preloaded.

##### ysql_catalog_preload_additional_tables

> When enabled, PG backend processes preloads the catalog tables: `pg_am`, `pg_amproc`, `pg_cast`, and `pg_tablespace`. This flag reduces latency of first query execution of a particular statement on a connection.

Reviewer 2: ```suggestion
When enabled, the postgres backend processes preload the `pg_am`, `pg_amproc`, `pg_cast`, and `pg_tablespace` catalog tables. This flag reduces latency of first query execution of a particular statement on a connection.
```

Default: `false`

If [ysql_catalog_preload_additional_table_list](#ysql-catalog-preload-additional-table-list) is also specified, the union of `pg_am`, `pg_amproc`, `pg_cast`, and `pg_tablespace` and the tables specified in `ysql_catalog_preload_additional_table_list` is preloaded.

##### ysql_enable_read_request_caching

Enables the YB-TServer catalog cache, which reduces YB-Master overhead for starting a connection and internal system catalog metadata refresh (for example, after executing a DDL), when there are many YSQL connections per node.

Default: `true`

##### ysql_use_relcache_file

> Controls whether to utilize the PostgreSQL relcache init file, which caches critical system catalog entries. If enabled, each PostgreSQL connection loads only this minimal set of cached entries (except if the relcache init file needs to be re-built, for example, after a DDL invalidates the cache). If disabled, each PostgreSQL connection preloads the catalog cache, which consumes more memory but reduces first query latency.

Reviewer 2: ```suggestion
Controls whether to use the PostgreSQL relcache init file, which caches critical system catalog entries. If enabled, each PostgreSQL connection loads only this minimal set of cached entries (except if the relcache init file needs to be re-built, for example, after a DDL invalidates the cache). If disabled, each PostgreSQL connection preloads the catalog cache, which consumes more memory but reduces first query latency.
```

Default: `true`

##### ysql_enable_db_catalog_version_mode

Enable the per database catalog version mode. A DDL statement that
affects the current database can only increment catalog version for
that database.

Default: `true`

{{< note title="Important" >}}

> Previously, after a DDL statement is executed, if the DDL statement increments the catalog

Reviewer 2: "Previously" - what do you mean? In versions prior to this one?

Author: @myang2021

Reviewer 3: @ddhodge @aishwarya24 , yes, we can change `Previously` to `In previous releases` if that's clearer.

version, then all the existing connections need to refresh catalog caches before
they execute the next statement. When per database catalog version mode is
enabled, multiple DDL statements can be concurrently executed if each DDL only
affects its current database and is executed in a separate database. Existing
connections only need to refresh their catalog caches if they are connected to
the same database as that of a DDL statement. It is recommended to keep the default value of this flag because per database catalog version mode helps to avoid unnecessary cross-database catalog cache refresh which is considered as an expensive operation.

{{< /note >}}

If you encounter any issues caused by per-database catalog version mode, you can disable per database catalog version mode using the following steps:

1. Shutdown the cluster.

1. Start the cluster with `--ysql_enable_db_catalog_version_mode=false`.

1. Execute the following YSQL statements:

    ```sql

>     SET yb_non_ddl_txn_for_sys_tables_allowed=true;

Reviewer 2: yb_non_ddl_txn_for_sys_tables_allowed and yb_fix_catalog_version_table not documented that I can see. These look like YSQL config parameters.

Author: cc @myang2021

Reviewer 3: `yb_non_ddl_txn_for_sys_tables_allowed` is an existing YB specific configuration parameter. I wasn't aware that it is not documented but I kind of understand why it is not documented. This is because it is not meant for every user to use because its usage is quite risky and if not used correctly it can lead to system errors. `yb_fix_catalog_version_table` is a new function introduced in this feature (DDL concurrency). For the same reason, we might now want to document it. From the commit message that added this function:

In per-database catalog version mode, the table `pg_yb_catalog_version` should
have exactly one row per database: no more, no less. This means that it is
important to keep the table `pg_database` and `pg_yb_catalog_version` in sync
so that they have exactly the same set of database oids. Likewise, in
global catalog version mode, the table `pg_yb_catalog_version` should only
have one row for database `template1`. Although it is possible to continue to
allow one row per database in global catalog version mode because we can only
make use of the row for database `template1`, it is still cleaner to delete
the extra rows that we are not going to use.

I added a new function `yb_fix_catalog_version_table` that can be used to fix
the `pg_yb_catalog_version` in case it is not in sync with the expected mode,
either due to code bug, or when cluster admin decides to manually convert the
cluster into the other mode.

(1) In global catalog version mode, `yb_fix_catalog_version_table` will delete
all the extra rows except for the row of database `template1`.
(2) In per-database catalog version mode, `yb_fix_catalog_version_table` will
ensure that `pg_yb_catalog_version` and `pg_database` stay in sync where the
table `pg_database` is considered as the source of truth. This is done by
inserting missing rows for db oids that exist in `pg_database` but not
in `pg_yb_catalog_version`, and delete extra rows for db oids that exist in
`pg_yb_catalog_version` but not in `pg_database`.

The new built-in function `yb_fix_catalog_version_table` can be written as:

```
CREATE OR REPLACE FUNCTION
  yb_fix_catalog_version_table(per_database_mode boolean)
RETURNS VOID
AS
$$
  INSERT INTO pg_catalog.pg_yb_catalog_version
    SELECT oid, 1, 1 FROM pg_catalog.pg_database
      WHERE per_database_mode AND
            (oid NOT IN (SELECT db_oid FROM pg_catalog.pg_yb_catalog_version));
  DELETE FROM pg_catalog.pg_yb_catalog_version
    WHERE (NOT per_database_mode AND db_oid != 1) OR
           (per_database_mode AND
            db_oid NOT IN (SELECT oid FROM pg_catalog.pg_database));
$$
LANGUAGE SQL STRICT;
REVOKE EXECUTE ON FUNCTION yb_fix_catalog_version_table(boolean) FROM public;
```

Author: Updated the code blocks @myang2021.

    SELECT yb_fix_catalog_version_table(false);
    ```

To re-enable the per database catalog version mode using the following steps:

1. Execute the following YSQL statements:

    ```sql
    SET yb_non_ddl_txn_for_sys_tables_allowed=true;
    SELECT yb_fix_catalog_version_table(true);
    ```

> 1. Shutdown the cluster.

Reviewer 2: ```suggestion
1. Shut down the cluster.
```

{{< note title="Important" >}}

Previously, after a DDL statement is executed, if the DDL statement increments the catalog
version, then all the existing connections need to refresh catalog caches before
they execute the next statement. When per database catalog version mode is
enabled, multiple DDL statements can be concurrently executed if each DDL only
affects its current database and is executed in a separate database. Existing
connections only need to refresh their catalog caches if they are connected to
the same database as that of a DDL statement. It is recommended to keep the default value of this flag because per database catalog version mode helps to avoid unnecessary cross-database catalog cache refresh which is considered as an expensive operation.

{{< /note >}}

If you encounter any issues caused by per-database catalog version mode, you can disable per database catalog version mode using the following steps:

> 1. Shutdown the cluster.

Reviewer 2: ```suggestion
1. Shut down the cluster.
```
